library/src/main/scala/scala/scalajs/runtime/RuntimeLong.scala

```diff
@@ -2,52 +2,153 @@ package scala.scalajs.runtime
-  def toInt: Int = l | (m << BITS)
-  def toLong: Long = x.asInstanceOf[Long]
+  // Binary compatibility for the old (l, m, h) encoding
+
+  @deprecated("Use the constructor with (lo, hi) instead.", "0.6.6")
+  def this(l: Int, m: Int, h: Int) =
+    this(l | (m << 22), (m >> 10) | (h << 12))
+
+  @deprecated("Use lo and hi instead.", "0.6.6")
+  def l: Int = lo & ((1 << 22) - 1)
+
+  @deprecated("Use lo and hi instead.", "0.6.6")
+  def m: Int = (lo >>> 22) & ((hi & ((1 << 12) - 1)) << 10)
+
+  @deprecated("Use lo and hi instead.", "0.6.6")
+  def h: Int = hi >>> 12
+
+  // Universal equality
+
+  override def equals(that: Any): Boolean = that match {
+    case b: RuntimeLong => inline_equals(b)
+    case _              => false
+  }
+
+  override def hashCode(): Int =
+    lo ^ hi
+
+  // String operations
+
+  override def toString(): String = {
+    val lo = this.lo
+    val hi = this.hi
+
+    if (isInt32(lo, hi)) {
+      lo.toString()
+    } else if (hi < 0) {
+      val (absLo, absHi) = inline_unary_-(lo, hi)
+      "-" + toUnsignedString(absLo, absHi)
+    } else {
+      toUnsignedString(lo, hi)
+    }
+  }
+
+  private def toUnsignedString(lo: Int, hi: Int): String = {
+    // This is called only if (lo, hi) is not an Int32
+
+    if (isUnsignedSafeDouble(hi)) {
+      // (lo, hi) is small enough to be a Double, use that directly
+      asUnsignedSafeDouble(lo, hi).toString
+    } else {
+      /* We divide (lo, hi) once by 10^9 and keep the remainder.
+       *
+       * The remainder must then be < 10^9, and is therefore an int32.
+       *
+       * The quotient must be <= ULong.MaxValue / 10^9, which is < 2^53, and
+       * is therefore a valid double. It must also be non-zero, since we tested
+       * previously for cases where (lo, hi) < 2^53, but 2^10 is itself < 2^53.
+       */
+      val TenPow9Lo = 1000000000L.toInt
+      val TenPow9Hi = (1000000000L >>> 32).toInt
+
+      val quotRem = unsignedDivModHelper(lo, hi, TenPow9Lo, TenPow9Hi,
+          AskBoth).asInstanceOf[js.Tuple4[Int, Int, Int, Int]]
+      val quotLo = quotRem._1
+      val quotHi = quotRem._2
+      val rem = quotRem._3 // remHi must be 0 by construction
+
+      val quot = asUnsignedSafeDouble(quotLo, quotHi)
+
+      val remStr = rem.toString
+      quot.toString + "000000000".jsSubstring(remStr.length) + remStr
+    }
+  }
+
+  // Conversions
+
+  def toByte: Byte = lo.toByte
+  def toShort: Short = lo.toShort
+  def toChar: Char = lo.toChar
+  def toInt: Int = lo
+  def toLong: Long = this.asInstanceOf[Long]
   def toFloat: Float = toDouble.toFloat
-  def toDouble: Double =
-    if (isMinValue) -9223372036854775808.0
-    else if (isNegative) -((-x).toDouble)
-    else l + m * TWO_PWR_22_DBL + h * TWO_PWR_44_DBL
+
+  def toDouble: Double = {
+    val lo = this.lo
+    val hi = this.hi
+
+    if (hi < 0) {
+      val (abslo, abshi) = inline_unary_-(lo, hi)
+      -(abshi.toUint * TwoPow32 + abslo.toUint) // abshi.toUint for MinValue
```

Comment:
In all other conversions to Double, we do `hi * TwoPow32 + lo.toUint`. But here we also need `hi.toUint` for the very lonely case where `hi == 0x80000000` and `lo == 0` (which is `Long.MinValue`), in which case `abshi` is still `0x80000000` but it must be interpreted as unsigned to give 2^31 (not -2^31).